keps/sig-scheduling/3022-min-domains-in-pod-topology-spread/README.md

```diff
@@ -0,0 +1,502 @@
+
+### Goals
+
+- Users can specify `minDomains` to limit the number of domains.
+- Users can use it as a mandatory requirement with `WhenUnsatisfiable=DoNotSchedule`.
+- Users can use it as a best-efforts manner with `WhenUnsatisfiable=ScheduleAnyway`.
+
+### Non-Goals
+
+- Add new field to limit the maximum number of topology domains.
+
+## Proposal
+
+### User Story
+
+I am using cluster autoscaler and I want to force spreading a deployment over at least 5 Nodes.
+
+## Design Details
+
+Users can define a minimum number of domains with `minDomains` parameter.
+
+Pod Topology Spread has the semantics of "global minimum", which means the minimum number of pods that match the label selector in a topology domain.
+
+When the number of domains that have matching Pods is less than `minDomains`,
+Pod Topology Spread treats "global minimum" as 0.
+
+As a result,
+
+- when `whenUnsatisfiable` equals `DoNotSchedule`, scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same or more number of matching Pods as `maxSkew`.
+- when `whenUnsatisfiable` equals `ScheduleAnyway`, scheduler prefers Nodes on the domains that have the less number of matching Pods as `maxSkew`.
+
+It is the optional parameter and the default value is 0.
+
+### API
+
+New optional parameter called `MinDomains` is introduced to `PodSpec.TopologySpreadConstraint[*]`.
+
+```go
+type TopologySpreadConstraint struct {
+......
+  // When the number of domains that have matching Pods is less than `minDomains`,
+  // Pod Topology Spread treats "global minimum" as 0.
+  // As a result,
+  // - when `whenUnsatisfiable` equals `DoNotSchedule`,
+  //   scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same number of matching Pods as `maxSkew`.
+  // - when `whenUnsatisfiable` equals `ScheduleAnyway`,
+  //   scheduler prefers Nodes on the domains that don't have the same number of matching Pods as `maxSkew`.
+  // Default value is 0.
+  // +optional
+  MinDomains *int32
+}
+```
+
+### Implementation details
+
+In Filter, filtering criteria is
+
+```
+('existing matching num' + 'if self-match (1 or 0)' - 'global min matching num') <= 'maxSkew'
+```
+
+- `existing matching num` denotes the number of current existing matching Pods on the domain.
+- `if self-match` denotes if the labels of Pod is match with selector of the constraint.
+- `global min matching num` denotes the minumun number of matching Pods.
+
+For `whenUnsatisfiable: DoNotSchedule`, Pod Topology Spread treats `global min matching num` as 0
+when the number of domains that have matching Pods is less than `minDomains`.
+
+And, in `preScoreState`, there is the `IgnoredNodes` field and Nodes in `preScoreState.IgnoredNodes` will literally be ignored and the score for those Nodes will be 0.
+
+For `whenUnsatisfiable: ScheduleAnyway`, Pod Topology Spread adds Nodes on the domains which have the same or more number of matching Pods as `maxSkew` to `preScoreState.IgnoredNodes`
```

Comment:
How do you calculate which nodes to ignore?

Reply:
We calculate the number of matching pods here. We can use this `cnt` variable to judge if the Node is preferred or not.
https://github.com/kubernetes/kubernetes/blob/108c284a330a82ce1a1f80238e4f54bf5e8b045a/pkg/scheduler/framework/plugins/podtopologyspread/scoring.go#L202-L208

But, now, I notice that we should not ignore the Node because even if the Node is not preferred from one constraint, the Node may be preferred from another constraint. In this case, the score of Node shouldn't be 0.
So, I'll fix this strategy. 🙇 

---

The scoring logic of Pod Topology Spread is not simple -- The higher the score from `Score` function, the lower will be the final score calculated by Normalized Score. Therefore, we can't use a simple strategy like adding 0 points [here](https://github.com/kubernetes/kubernetes/blob/108c284a330a82ce1a1f80238e4f54bf5e8b045a/pkg/scheduler/framework/plugins/podtopologyspread/scoring.go#L209) if the Node is not prefered for the constraint.

In `Score` function, scores for each constraint are calculated in this [function](https://github.com/kubernetes/kubernetes/blob/108c284a330a82ce1a1f80238e4f54bf5e8b045a/pkg/scheduler/framework/plugins/podtopologyspread/scoring.go#L293). 
Since the values of `MaxSkew` and `tpWeight` are fixed for each constraint, the larger `cnt`, the larger the score. And as mentioned, in Pod Topology Spread, the higher your score, the lower your final score will be. 

So, I think we need to calculate the global maximum matching num for each constraint in PreScore so that we can give non-preferred Node the lowest score finally among the scores calculated by that constraint.

@Huang-Wei What do you think about this logic? Please correct me if I'm understanding something incorrectly.

I know that my written English may be difficult to understand. If my explanation is not clear, please let me know. 🙇

Reply:
I found that I hadn't thought about it enough, so I revised the whole implementation detail of score phase. 🙏